### ISSUE TITLE
Delay rebuild?

### ISSUE THREAD

**Issue by afcowie** [digital-asset/ghcide#302]
Might it be possible to put a delay (configurable, perhaps) of some short time (say, a couple seconds) before firing off the recompile? I'm used to **ghcid** reacting when I _save_ a file, but having the child GHC process maxing out the CPUs when I'm still in the middle of typing seems a bit overzealous?

AfC

**Comment by afcowie** [comment]
I realized that I would be pretty happy with only rebuilding on error, so set out to see if I could add such a feature myself. I've been looking around in the **ghcide** code base and couldn't quite figure out what was triggering the shake builds. If someone could point me to which event (ie, it's not "CodeAction" or "Completions" but....) I'd be happy to try and run up a patch.

**Comment by cocreature** [comment]
I don’t really want to change the current behavior in ghcide. If users want a delay this should imho be on the client, i.e. the editor side by limiting the rate at which it sends document updates. As for the code that triggers the updates this is in https://github.com/digital-asset/ghcide/blob/master/src/Development/IDE/LSP/Notifications.hs#L42.

**Comment by afcowie** [comment]
@cocreature fair enough. So this is something that should be changed in the VS Code **ghcide** extension? (I admit I am overlapping the two conceptually)

Firing off a GHC build—and then cancelling and restarting—every keystroke is making my kernel think it is busy and really ramping the CPU up out of low power mode, resulting in massive laptop battery drain. I had to switch to **vim** and **ghcid** while commuting just so my machine would last!

**Comment by ndmitchell** [comment]
If it works nicely in ghcid, maybe what you want is only checking on save? I think that's something that is in a few extensions, e.g. https://github.com/DavidAnson/vscode-markdownlint/issues/28

**Comment by domenkozar** [comment]
Most VSCode extensions offer an option to pick between `instant`, `save` or `never`.

**Comment by afcowie** [comment]
@ndmitchell yeah, checking on save would be great. Changed title. Is this the right repo to have this issue open in?

**Comment by afcowie** [comment]
(certainly a configuration setting like @domenkozar mentions would be a VS Code specific question, but does **ghcide** need to change to accommodate this?)

**Comment by ndmitchell** [comment]
Ghcide repo includes the VS Code extension, so yes, it does. It probably also wants to be an option to the Haskell piece, so in the core.

**Comment by jneira** [comment]
This seems to belongs to the vscode extension so maybe should be transferred to vscode-haskell
